src/analysis/complex/upper_half_plane.lean

```diff
@@ -0,0 +1,165 @@
+/-
+Copyright (c) 2021 Alex Kontorovich and Heather Macbeth and Marc Masdeu. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Alex Kontorovich, Heather Macbeth, Marc Masdeu
+-/
+
+import linear_algebra.special_linear_group
+import analysis.complex.basic
+import group_theory.group_action.defs
+
+/-!
+# The upper half plane and its automorphisms
+
+This file defines `upper_half_plane` to be the upper half plane in `ℂ`.
+
+We furthermore equip it with the structure of an `SL(2,ℝ)` action by
+fractional linear transformations.
+
+We define the notation `ℍ` for the upper half plane available in the locale
+`upper_half_plane` so as not to conflict with the quaternions.
+-/
+
+noncomputable theory
+
+open matrix matrix.special_linear_group
+
+open_locale classical big_operators matrix_groups
+
+local attribute [instance] fintype.card_fin_even
+
+/-- The upper half plane -/
+abbreviation upper_half_plane :=
+{point : ℂ // 0 < point.im}
+
+localized "notation `ℍ` := upper_half_plane" in upper_half_plane
+
+namespace upper_half_plane
+
+/-- Imaginary part -/
+def im (z : ℍ) := (z : ℂ).im
+
+/-- Real part -/
+def re (z : ℍ) := (z : ℂ).re
+
+@[simp] lemma coe_im (z : ℍ) : (z : ℂ).im = z.im := rfl
+
+@[simp] lemma coe_re (z : ℍ) : (z : ℂ).re = z.re := rfl
+
+lemma im_pos (z : ℍ) : 0 < z.im := z.2
+
+lemma im_ne_zero (z : ℍ) : z.im ≠ 0 := z.im_pos.ne'
+
+lemma ne_zero (z : ℍ) : (z : ℂ) ≠ 0 :=
+begin
+  intros h,
+  apply z.im_ne_zero,
+  rw [←complex.zero_im, ←h, coe_im]
+end
+
+lemma norm_sq_pos (z : ℍ) : 0 < complex.norm_sq (z : ℂ) :=
+begin
+  rw complex.norm_sq_pos,
+  exact z.ne_zero
+end
+
+lemma norm_sq_ne_zero (z : ℍ) : complex.norm_sq (z : ℂ) ≠ 0 := (norm_sq_pos z).ne'
+
+/-- Numerator of the formula for a fractional linear transformation -/
+@[simp] def top (g : SL(2, ℝ)) (z : ℍ) : ℂ := (g 0 0) * z + (g 0 1)
+
+/-- Denominator of the formula for a fractional linear transformation -/
+@[simp] def bottom (g : SL(2, ℝ)) (z : ℍ) : ℂ := (g 1 0) * z + (g 1 1)
```

Comment:
Does it make sense to call these `num` and `denom`? Just like we have `rat.num` and `rat.denom`.
Also, I'm not sure whether these should be in `upper_half_plane`, but I don't have a better suggestion atm.

It is quite uncommon to mark definitions as `@[simp]`. Are you sure you want these definitions to always be simped away? You can add `local attribute [simp] top bottom` below these definitions if you want that behaviour in this section.

Reply:
Will change top/bottom to num/denom, but we've found that we mostly do want simp to clear top/bottom because it helps for the arguments in this file and the next PR. On the rare occasion that we want facts about denominators being nonzero, we add [-bottom] to keep simp from unfolding the definition.